PingWorker not work

Work [ id=42e1814b-15be-45aa-9fba-293602b236b0, tags={ info.mqtt.android.service.ping.PingWorker } ] was cancelled
                 java.util.concurrent.CancellationException: Task was cancelled.
                 	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1184)
                 	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:514)
                 	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
                 	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:317)
                 	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
                 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
                 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
                 	at java.lang.Thread.run(Thread.java:1012)

since 4.1 i have noticed this as well but it makes it so that the connection has more losses then i had with 3.5.1

2023-09-29 09:06:22.148 28065-28106 WM-WorkerWrapper        com.qbus.ubie                        I  Work [ id=3e113b43-0d60-4a35-8d38-2e20860e8a54, tags={ info.mqtt.android.service.ping.PingWorker } ] was cancelled
                                                                                                    java.util.concurrent.CancellationException: Task was cancelled.
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1184)
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:514)
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
                                                                                                    	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:317)
                                                                                                    	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
                                                                                                    	at java.lang.Thread.run(Thread.java:923)
2023-09-29 09:06:40.685 28065-28079 com.qbus.ubie           com.qbus.ubie                        I  Background concurrent copying GC freed 80343(6260KB) AllocSpace objects, 0(0B) LOS objects, 46% free, 6950KB/12MB, paused 159us total 113.095ms
2023-09-29 09:06:48.205 28065-28662 ContentValues           com.qbus.ubie                        I  reveived packet: UBIE?
2023-09-29 09:06:48.491 28065-28662 ContentValues           com.qbus.ubie                        I  reveived packet: UBIE:UB10000005EE:192.168.75.120:255.255.255.0:192.168.75.1:192.168.75.3:3000
2023-09-29 09:06:48.808 28065-28662 ContentValues           com.qbus.ubie                        I  reveived packet: UBIE:PROTO0000009:192.168.75.110:255.255.255.0:192.168.75.1:192.168.75.3:3000
2023-09-29 09:06:57.704 28065-28079 com.qbus.ubie           com.qbus.ubie                        I  Background young concurrent copying GC freed 72797(5959KB) AllocSpace objects, 0(0B) LOS objects, 45% free, 7161KB/12MB, paused 5.371ms total 52.301ms
2023-09-29 09:07:18.205 28065-28662 ContentValues           com.qbus.ubie                        I  reveived packet: UBIE?
2023-09-29 09:07:18.497 28065-28662 ContentValues           com.qbus.ubie                        I  reveived packet: UBIE:UB10000005EE:192.168.75.120:255.255.255.0:192.168.75.1:192.168.75.3:3000
2023-09-29 09:07:18.802 28065-28662 ContentValues           com.qbus.ubie                        I  reveived packet: UBIE:PROTO0000009:192.168.75.110:255.255.255.0:192.168.75.1:192.168.75.3:3000
2023-09-29 09:07:22.796 28065-20168 mqtt                    com.qbus.ubie                        E  connection lost cloud
                                                                                                    Connection lost (32109) - java.io.EOFException
                                                                                                    	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:197)
                                                                                                    	at java.lang.Thread.run(Thread.java:923)
                                                                                                    Caused by: java.io.EOFException
                                                                                                    	at java.io.DataInputStream.readByte(DataInputStream.java:270)
                                                                                                    	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:92)
                                                                                                    	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:137)
                                                                                                    	at java.lang.Thread.run(Thread.java:923) 
2023-09-29 09:07:22.964 28065-28665 WM-Processor            com.qbus.ubie                        W  Didn't find WorkSpec for id WorkGenerationalId(workSpecId=263da75c-44d6-4b8a-b408-5a08fa7a3557, generation=0)
2023-09-29 09:07:23.802 28065-20307 System.out              com.qbus.ubie                        I  [socket]:check permission begin!
2023-09-29 09:07:24.301 28065-20014 mqtt                    com.qbus.ubie                        I  connected to mqtt, was reconnect: true

--- a/serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt
+++ b/serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt
@@ -53,21 +53,31 @@ internal class AlarmPingSender(
     override fun schedule(delayInMilliseconds: Long) {
         Timber.d("$id: Schedule next alarm at ${sdf.format(Date(System.currentTimeMillis() + delayInMilliseconds))}")
 
-        val pingWork = OneTimeWorkRequest.Builder(PingWorker::class.java)
-        val data = Data.Builder()
-        data.putBoolean("logging", pingLogging)
-        data.putInt("keepRecordCount", keepPingRecords)
-        data.putString("id", id)
+        // Check if work is already scheduled with this ID
+        val workInfo = workManager.getWorkInfosForUniqueWork("${PING_JOB}_$id").get()
 
-        pingWork
-            .setInitialDelay(delayInMilliseconds, TimeUnit.MILLISECONDS)
-            .setInputData(data.build())
+        // Only schedule if no work exists or all existing work is finished/cancelled
+        if (workInfo.isEmpty() || workInfo.all { it.state.isFinished }) {
+            val pingWork = OneTimeWorkRequest.Builder(PingWorker::class.java)
+            val data = Data.Builder()
+            data.putBoolean("logging", pingLogging)
+            data.putInt("keepRecordCount", keepPingRecords)
+            data.putString("id", id)
 
-        workManager.enqueueUniqueWork(
-            "${PING_JOB}_$id",
-            ExistingWorkPolicy.REPLACE,
-            pingWork.build()
-        )
+            pingWork
+                .setInitialDelay(30, TimeUnit.SECONDS)
+                .setInputData(data.build())
+
+            workManager.enqueueUniqueWork(
+                "${PING_JOB}_$id",
+                ExistingWorkPolicy.REPLACE,
+                pingWork.build()
+            )
+
+            Timber.d("$id: Successfully scheduled new ping job")
+        } else {
+            Timber.w("$id: I GOT YOU Ping job already scheduled, skipping")
+        }
     }
 
     companion object {

--- a/serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt
+++ b/serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt
@@ -65,7 +65,8 @@ internal class AlarmPingSender(
             data.putString("id", id)
 
             pingWork
-                .setInitialDelay(30, TimeUnit.SECONDS)
+                .setInitialDelay(delayInMilliseconds, TimeUnit.MILLISECONDS)
+//                .setInitialDelay(30, TimeUnit.SECONDS)
                 .setInputData(data.build())
 
             workManager.enqueueUniqueWork(

